core/state/statedb.go

```diff
@@ -652,27 +652,24 @@ func (s *StateDB) setStateObject(object *stateObject) {
 
 // GetOrNewStateObject retrieves a state object or create a new state object if nil.
 func (s *StateDB) GetOrNewStateObject(addr common.Address) *stateObject {
+	return s.getOrNewStateObject(addr, false)
+}
+
+func (s *StateDB) getOrNewStateObject(addr common.Address, checkPrecompile bool) *stateObject {
 	stateObject := s.getStateObject(addr)
 	if stateObject == nil {
-		stateObject, _ = s.createObject(addr)
+		stateObject, _ = s.createObject(addr, checkPrecompile)
 	}
 	return stateObject
 }
 
 // createObject creates a new state object. If there is an existing account with
 // the given address, it is overwritten and returned as the second return value.
-func (s *StateDB) createObject(addr common.Address) (newobj, prev *stateObject) {
+func (s *StateDB) createObject(addr common.Address, checkPrecompile bool) (newobj, prev *stateObject) {
 	prev = s.getDeletedStateObject(addr) // Note, prev might have been deleted, we need that!
 	newobj = newObject(s, addr, nil)
 	if prev == nil {
 		s.journal.append(createObjectChange{account: &addr})
-		if s.logger != nil {
-			// Precompiled contracts are touched during a call.
-			// Make sure we avoid emitting a new account event for them.
-			if _, ok := s.precompiles[addr]; !ok {
-				s.logger.OnNewAccount(addr)
```

Comment:
Nah not really. The thing here is that all this code is actually there to match Firehose legacy tracer which is the one that is wrong.

The code is ok. It sadden me that we need to tweak the Geth code so much to get back the old behavior. I'm trying to think of alternative implementations now. It seems that on a single occasion we set `checkPrecompile` at `true` and it's in the `AddBalance` of a `StaticCall`.

Tomorrow morning I'll work on adding to battlefield some test cases to cover those details and then we could try to come up with something "less" intrusive. 

But yeah I do see why we need to tweak the `StateDB` tracing code since the event is not emitted at all, we have little way to get notified. Something I'm thinking about, we remove the passing of the `checkPrecompile` everywhere and we always emit in our version (with probably some kind of variable to enable this behavior, maybe at the EVM/StateDB level).

Then we deal with respecting the Firehose legacy behavior in `OnNewAccount` itself since now we would always receive the event.

Reply:
Does this mean that this with PR, we are now on par with Firehose legacy? In which range have you completed the validation so far for Geth?

Reply:
@maoueh I wasn't referring to the check compile logic here. 
I meant `s.logger.OnNewAccount(addr)` should be outside the `prev == nil` condition.

Reply:
We have tested till 3.5M. I did find the same `Topics: [""]` mismatch issues in this as well.

Reply:
> I wasn't referring to the checkPrecompile

Ok sorry, misunderstood indeed. This is also something that was wrong in the old Firehose implementation, we discussed me and the Geth team member about this and the told me our old instrumentation point was incorrect.

I'll merge and work on your PR. 

> I did find the same Topics: [""]

Perfect, at least the Geth and Erigon are equivalent, which is a good news.